### ISSUE TITLE
Bug for download filled map.

### ISSUE THREAD

**Issue by shved255** [PrismarineJS/mineflayer#3487]
Hello everyone, please help, I have a problem with the filled map. My bot logs into the server, I instruct it to download the image from the map that is in the frame. It seems that it is possible to pump out, but all the cards are turned over and downloaded randomly. If anyone knows, please help how to download images from the filled map normally, not upside down? And how can you implement the download of maps so that they are not downloaded randomly, for example, from left to right? Probably a bug of MineFlayer itself, if not, sorry. I've tried using the item-map-downloader plugin, but it's still the same. I went through many forums, did not find an answer. On github, I found one code that can be used to download maps. I edited it a little, I also tried and it didn't work out to download the map normally, here is his link: https://pastebin.com/jE3KF5XL I will be very grateful for your help, thank you. I tried using different versions of the bot, it didn't work either. Image filled maps: 

![изображение](https://github.com/user-attachments/assets/65ff92c4-7bc6-4685-8bf6-be9c92ac5628)

**Comment by rom1504** [comment]
Send errors and bugs

**Comment by shved255** [comment]
![map_0](https://github.com/user-attachments/assets/2adefa9f-408e-4c98-a4e7-178d76e6d635)
![map_1](https://github.com/user-attachments/assets/57dec61a-55a7-4a57-930a-5e87b7864349)
![map_2](https://github.com/user-attachments/assets/f4dda1fd-2bd1-4cf8-be16-1f95ff2e9dde)
![map_3](https://github.com/user-attachments/assets/7891e114-1870-4a8b-8ecf-137016fff7a9)
![map_4](https://github.com/user-attachments/assets/9c99aae9-ba38-49b7-97bb-60489b90e156)
![map_5](https://github.com/user-attachments/assets/6b27840c-b05d-4ca4-a2eb-4903d53b7b19)
![map_6](https://github.com/user-attachments/assets/3806514f-9992-49b4-8630-3c64bffc7290)
![map_7](https://github.com/user-attachments/assets/3cadbcf7-5f4e-4835-94ed-899925e60e48)
![map_8](https://github.com/user-attachments/assets/53c42d94-58c8-47fe-ab16-ade460d11cad)
![map_9](https://github.com/user-attachments/assets/93cd180c-3a82-42ba-a972-91b34be6b798)
![map_10](https://github.com/user-attachments/assets/6eaad291-e0b4-4d43-b3a5-c73060c795c1)
![map_11](https://github.com/user-attachments/assets/c725e8a1-7c70-4e30-95ba-5bdeda612b57)


я пытался их объеденить через canvas, получался бред. Я запускал бота в локальный мир и ставил filled map в рамку, и было перевернутое изображение.

**Comment by shved255** [comment]
Please help, I really need a bug fix

**Comment by rom1504** [comment]
https://github.com/PrismarineJS/minecraft-data/blob/master/data/pc/latest/proto.yml#L1878

This is the definition of this packet

You need the use the other fields, not only data

**Comment by rom1504** [comment]
Also see https://wiki.vg/Protocol#Map_Data

This packet updates a part of a map. It is sent many times, you need to all the values to assemble the image.

**Comment by rom1504** [comment]
Would welcome a PR implementing this directly in mineflayer

**Comment by shved255** [comment]
thanks, but I don't understand how to use it. Maybe there are examples?

**Comment by rom1504** [comment]
There's nothing to use, new code needs to be written.

You can try it out, maybe you can ask chatgpt for help (giving it existing code and API doc as context)

**Comment by shved255** [comment]
Thank you very much, I'll try, I'll write what yes

**Comment by shved255** [comment]
Thats workd! Thanks you veru much. But how do you get the coordinates of the frame whose data you received? I used MapData and x, y, z methods, also used PosY, PosX, writes for instead of the first case 1 in all maps, in the other one completely undefined. Please tell me how to get the coordinates of the frame, and whether there is such a thing at all. I use the event bot._client.on('map'....

**Comment by rom1504** [comment]
You need to keep track using the map id

**Comment by shved255** [comment]
Of course, I'm embarrassed to ask, but what functions or methods should I use? The neural network does not help me ;(

**Comment by shved255** [comment]
Without a coordinate function, you get something like this:
![combined_map](https://github.com/user-attachments/assets/d6b06134-5a8f-4484-8d2d-9ee3ea1c3643)

**Comment by rom1504** [comment]
Can you log the map packets you get (without the data field) and send them here ?
It'll be easier to help that way

**Comment by shved255** [comment]
Thats map data:


  itemDamage: 83,
  scale: 0,
  locked: false,
  icons: undefined,
  columns: 128,
  rows: 128,
  x: 0,
  y: 0,
  data: <Buffer 22 3a 22 3a 22 3a 22 3a 22 3a 22 3a 22 3a 22 3a 22 3a 22 3a 22 3a 22 3a 22 3a 22 3a 22 3a 22 3a 22 3a 22 3a 22 3a 22 3a 22 3a 22 3a 22 3a 22 3a 22 3a ... 16334 more bytes>

**Comment by shved255** [comment]
but all maps are x and y this 0

**Comment by rom1504** [comment]
Can you send the other ones

**Comment by shved255** [comment]
yes, this code on English:

https://pastebin.com/p80WJZUP

**Comment by shved255** [comment]
Sorry for the long wait

**Comment by rom1504** [comment]
No I am asking for the other map packets logs

**Comment by shved255** [comment]
I didn't understand what kind of data, it seems that's all there is.

**Comment by shved255** [comment]
I got it, I'm going to do it now

**Comment by rom1504** [comment]
This seems to contain every packet.

What I am asking is all the `map` packets you receive

**Comment by shved255** [comment]
[map_logs.txt](https://github.com/user-attachments/files/17561602/map_logs.txt)

Most likely it is, or I do not know which logs to issue. I just started working with mindflayer recently, sorry.

**Comment by rom1504** [comment]
Would be easier to read without the data field.

But anyway that confirms what I was thinking.

"ItemDamage" field is what contains the coordinates.

We need to parse it better

**Comment by shved255** [comment]
and how to analyze it? item Damage has only 1 digit, but the coordinate axes are 3.

**Comment by shved255** [comment]
x=64+128×(itemDamage)
z=64+128×(itemDamage​)

should the formulas work?

**Comment by rom1504** [comment]
No very unlikely.

We need to check the decompiled code

**Comment by rom1504** [comment]
https://github.com/IceTank/mineflayer-item-map-downloader/blob/master/index.js#L68 seems to think this damage field is the map id.

Maybe the position information is in entity metadata

**Comment by rom1504** [comment]
Yeah looking at your log above, I believe the information is present in objectData field. Will need a bit of code to piece it all together

**Comment by shved255** [comment]
thanks, I'll try this.

**Comment by shved255** [comment]
of course, I also noticed that map Data X and Y are empty only on verification. And with the help of item Damage, random numbers just appear, that is, they are similar, but in fact this is nonsense, for example, I have X = 0, Y = 2, the next card is X = 64, Y = -128.

**Comment by shved255** [comment]
Here is an example of maps that you could find on the server with a poison bouquet and a selection:

Received map data: {
Element damage: 0,
Scale: 0,
Tracking position: true,
blocked: false,
icons: [ { type: 1, x: -63, z: -93, direction: 4, display name: undefined } ],
columns: -128,
rows: -128,
x: 0,
y: 0,
Data: <Buffer 36 09 09 09 09 09 09 09 09 09 09 09 09 09 09 09 09 09 09 09 09 09 09 09 09 09 09 09 09 09 09 09 09 09 09 09 09 09 09 09 09 09 09 09 09 09 09 09 09 09 ... another 16334 bytes>
}

X and Z has numbers. Although the coordinates are in icons, they are also true. If you go to the verification server, then ICONS will be zero there.

**Comment by GefestM** [comment]
Have you solved this somehow? If yes, write to me

**Comment by shved255** [comment]
https://github.com/ImPhoba/FlayerCaptcha

> Have you solved this somehow? If yes, write to me

i solved this by FlayerCaptcha plugin.
